File: ios/Interactable/InteractableView.m
//
//  InteractableView.m
//  Interactable
//
//  Created by Tal Kol on 1/27/17.
//  Copyright © 2017 Wix. All rights reserved.
//

#import "InteractableView.h"
#import <React/UIView+React.h>
#import <React/RCTRootView.h>

@interface InteractableView()
@property (nonatomic, assign) BOOL originSet;
@property (nonatomic, assign) CGPoint origin;
@property (nonatomic) PhysicsAnimator *animator;
@property (nonatomic) PhysicsBehavior *dragBehavior;
@property (nonatomic, assign) CGPoint dragStartCenter;
@property (nonatomic, assign) CGPoint dragStartLocation;
@property (nonatomic, assign) BOOL initialPositionSet;
@property (nonatomic, assign) BOOL reactRelayoutHappening;
@property (nonatomic, assign) CGPoint reactRelayoutCenterDeltaFromOrigin;
@end

@implementation InteractableView

- (instancetype)init
{
    if ((self = [super init]))
    {
        self.originSet = NO;
        self.initialPositionSet = NO;
        self.reactRelayoutHappening = NO;
        
        // pan gesture recognizer for touches
        UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(handlePan:)];
        pan.delegate = self;
        [self addGestureRecognizer:pan];
    }
    return self;
}

- (void)reactSetFrame:(CGRect)frame
{
    // to handle the react relayout we need to remember the delta from previous origin
    self.reactRelayoutCenterDeltaFromOrigin = CGPointMake(self.origin.x - self.center.x, self.origin.y - self.center.y);
    
    self.origin = CGPointMake(CGRectGetMidX(frame), CGRectGetMidY(frame));
    self.originSet = YES;
    
    self.reactRelayoutHappening = YES;
    [super reactSetFrame:frame];
    self.reactRelayoutHappening = NO;
    
    // initial position
    if (!self.initialPositionSet)
    {
        self.initialPositionSet = YES;
        if (!CGPointEqualToPoint(self.initialPosition, CGPointZero))
        {
            self.center = CGPointMake(self.origin.x + self.initialPosition.x, self.origin.y + self.initialPosition.y);
        }
        
        // make sure this is after origin is set and happens once
        [self initializeAnimator];
    }
}

- (void)setCenter:(CGPoint)center
{
    if (self.initialPositionSet && self.reactRelayoutHappening)
    {
        // to handle a react relayout we maintain the same delta but now with the new origin
        center = CGPointMake(self.origin.x - self.reactRelayoutCenterDeltaFromOrigin.x, self.origin.y - self.reactRelayoutCenterDeltaFromOrigin.y);
    }
    
    if (self.originSet)
    {
        if (self.horizontalOnly) center.y = self.origin.y;
        if (self.verticalOnly) center.x = self.origin.x;
        
        if (self.limitX)
        {
            if (center.x - self.origin.x < self.limitX.min) center.x = self.limitX.min + self.origin.x;
            if (center.x - self.origin.x > self.limitX.max) center.x = self.limitX.max + self.origin.x;
        }
        if (self.limitY)
        {
            if (center.y - self.origin.y < self.limitY.min) center.y = self.limitY.min + self.origin.y;
            if (center.y - self.origin.y > self.limitY.max) center.y = self.limitY.max + self.origin.y;
        }
    }
  
    [super setCenter:center];
    [self reportAnimatedEvent];
}

- (void)initializeAnimator
{
    self.animator = [[PhysicsAnimator alloc] init];
    self.animator.delegate = self;
    
    // initialize constant behaviors
    if (self.springPoints)
    {
        for (InteractablePoint *point in self.springPoints) [self addConstantSpringBehavior:point];
    }
    if (self.gravityPoints)
    {
        for (InteractablePoint *point in self.gravityPoints) [self addConstantGravityBehavior:point];
    }
    if (self.frictionAreas)
    {
        for (InteractablePoint *point in self.frictionAreas) [self addConstantFrictionBehavior:point];
    }
}

- (void)physicsAnimatorDidPause:(PhysicsAnimator *)animator
{
    if (self.onSnap)
    {
        InteractablePoint *snapPoint = [InteractablePoint findClosestPoint:self.snapPoints toPoint:self.center withOrigin:self.origin];
        if (snapPoint)
        {
            self.onSnap(@
            {
                @"index": @([self.snapPoints indexOfObject:snapPoint]),
                @"id": snapPoint.id
            });
        }
    }
    
    if (self.onStop)
    {
        CGPoint deltaFromOrigin = [InteractablePoint deltaBetweenPoint:self.center andOrigin:self.origin];
        self.onStop(@
        {
            @"x": @(deltaFromOrigin.x),
            @"y": @(deltaFromOrigin.y)
        });
    }
}

- (void)reportAnimatedEvent
{
    if (self.onAnimatedEvent && self.originSet)
    {
        CGPoint deltaFromOrigin = [InteractablePoint deltaBetweenPoint:self.center andOrigin:self.origin];
        self.onAnimatedEvent(@
        {
            @"x": @(deltaFromOrigin.x),
            @"y": @(deltaFromOrigin.y)
        });
    }
}

// MARK: - Touches

- (void)handlePan:(UIPanGestureRecognizer *)pan
{
    if (pan.state == UIGestureRecognizerStateBegan)
    {
        [self cancelCurrentReactTouch];
        self.dragStartCenter = self.center;
        [self setTempBehaviorsForDragStart];
    }
    
    CGPoint translation = [pan translationInView:self];
    self.dragBehavior.anchorPoint = CGPointMake(self.dragStartCenter.x + translation.x, self.dragStartCenter.y + translation.y);
    [self.animator ensureRunning];
    
    if (pan.state == UIGestureRecognizerStateEnded ||
        pan.state == UIGestureRecognizerStateFailed ||
        pan.state == UIGestureRecognizerStateCancelled)
    {
        [self setTempBehaviorsForDragEnd];
    }
}

- (BOOL)gestureRecognizerShouldBegin:(UIPanGestureRecognizer *)pan
{
    CGPoint translation = [pan translationInView:self];
    if (self.horizontalOnly) return fabs(translation.x) > fabs(translation.y);
    if (self.verticalOnly) return fabs(translation.y) > fabs(translation.x);
    return YES;
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)pan shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)other
{
    // return YES to allow a hosting scrollview to scroll while an interactable view is moving
    return NO;
}

- (void)cancelCurrentReactTouch
{
    RCTRootView *view = [self getRootView];
    if (view != nil)
    {
        [(RCTRootView*)view cancelTouches];
    }
}

- (RCTRootView*)getRootView
{
    UIView *view = self;
    while (view.superview != nil)
    {
        view = view.superview;
        if ([view isKindOfClass:[RCTRootView class]]) break;
    }
    
    if ([view isKindOfClass:[RCTRootView class]])
    {
        return view;
    }
    return nil;
}

/*
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [[event allTouches] anyObject];
    self.dragStartCenter = self.center;
    self.dragStartLocation = [touch locationInView:self.superview];
    
    [self setTempBehaviorsForDragStart];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [[event allTouches] anyObject];
    CGPoint location = [touch locationInView:self.superview];
    CGFloat newCenterX = self.dragStartCenter.x + location.x - self.dragStartLocation.x;
    CGFloat newCenterY = self.dragStartCenter.y + location.y - self.dragStartLocation.y;
    
    self.dragBehavior.anchorPoint = CGPointMake(newCenterX, newCenterY);
    [self.animator ensureRunning];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [self setTempBehaviorsForDragEnd];
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [self setTempBehaviorsForDragEnd];
}
*/
 
- (void)setTempBehaviorsForDragStart
{
    [self.animator removeTempBehaviors];
    self.dragBehavior = nil;
    
    self.dragBehavior = [self addTempDragBehavior:self.drag];
}

- (void)setTempBehaviorsForDragEnd
{
    [self.animator removeTempBehaviors];
    self.dragBehavior = nil;
    
    CGPoint velocity = [self.animator getTargetVelocity:self];
    if (self.horizontalOnly) velocity.y = 0;
    if (self.verticalOnly) velocity.x = 0;
    CGFloat toss = 0.1;
    if (self.drag) toss = self.drag.toss;
    CGPoint projectedCenter = CGPointMake(self.center.x + toss*velocity.x, self.center.y + toss*velocity.y);
    
    InteractablePoint *snapPoint = [InteractablePoint findClosestPoint:self.snapPoints toPoint:projectedCenter withOrigin:self.origin];
    if (snapPoint) [self addTempSnapToPointBehavior:snapPoint];
    
    [self addTempBounceBehaviorWithLimitX:self.limitX limitY:self.limitY];
}

// MARK: - Behaviors

- (PhysicsBehavior*)addTempDragBehavior:(InteractableDrag*)drag
{
    PhysicsBehavior *res = nil;
    
    if (!drag || drag.tension == CGFLOAT_MAX)
    {
        PhysicsAnchorBehavior *anchorBehavior = [[PhysicsAnchorBehavior alloc] initWithTarget:self anchorPoint:self.center];
        res = anchorBehavior;
        [self.animator addTempBehavior:anchorBehavior];
    }
    else
    {
        PhysicsSpringBehavior *springBehavior = [[PhysicsSpringBehavior alloc] initWithTarget:self anchorPoint:self.center];
        springBehavior.tension = drag.tension;
        res = springBehavior;
        [self.animator addTempBehavior:springBehavior];
    }
    
    if (drag && drag.damping > 0.0)
    {
        PhysicsFrictionBehavior *frictionBehavior = [[PhysicsFrictionBehavior alloc] initWithTarget:self];
        frictionBehavior.friction = drag.damping;
        [self.animator addTempBehavior:frictionBehavior];
    }
    
    return res;
}

- (void)addTempSnapToPointBehavior:(InteractablePoint*)snapPoint
{
    PhysicsSpringBehavior *snapBehavior = [[PhysicsSpringBehavior alloc] initWithTarget:self anchorPoint:[snapPoint positionWithOrigin:self.origin]];
    snapBehavior.tension = snapPoint.tension;
    [self.animator addTempBehavior:snapBehavior];
    
    CGFloat damping = 0.7;
    if (snapPoint.damping > 0.0) damping = snapPoint.damping;
    PhysicsFrictionBehavior *frictionBehavior = [[PhysicsFrictionBehavior alloc] initWithTarget:self];
    frictionBehavior.friction = damping;
    [self.animator addTempBehavior:frictionBehavior];
}

- (void)addTempBounceBehaviorWithLimitX:(InteractableLimit*)limitX limitY:(InteractableLimit*)limitY
{
    if (limitX && limitX.bounce > 0.0)
    {
        CGPoint minPoint = CGPointMake(-CGFLOAT_MAX, -CGFLOAT_MAX);
        if (limitX.min != -CGFLOAT_MAX) minPoint.x = self.origin.x + limitX.min;
        CGPoint maxPoint = CGPointMake(CGFLOAT_MAX, CGFLOAT_MAX);
        if (limitX.max != CGFLOAT_MAX) maxPoint.x = self.origin.x + limitX.max;
        PhysicsBounceBehavior *bounceBehavior = [[PhysicsBounceBehavior alloc] initWithTarget:self minPoint:minPoint maxPoint:maxPoint];
        bounceBehavior.bounce = limitX.bounce;
        bounceBehavior.haptics = limitX.haptics;
        [self.animator addTempBehavior:bounceBehavior];
    }
    if (limitY && limitY.bounce > 0.0)
    {
        CGPoint minPoint = CGPointMake(-CGFLOAT_MAX, -CGFLOAT_MAX);
        if (limitY.min != -CGFLOAT_MAX) minPoint.y = self.origin.y + limitY.min;
        CGPoint maxPoint = CGPointMake(CGFLOAT_MAX, CGFLOAT_MAX);
        if (limitY.max != CGFLOAT_MAX) maxPoint.y = self.origin.y + limitY.max;
        PhysicsBounceBehavior *bounceBehavior = [[PhysicsBounceBehavior alloc] initWithTarget:self minPoint:minPoint maxPoint:maxPoint];
        bounceBehavior.bounce = limitY.bounce;
        bounceBehavior.haptics = limitY.haptics;
        [self.animator addTempBehavior:bounceBehavior];
    }
}

- (void)addConstantSpringBehavior:(InteractablePoint*)point
{
    CGPoint anchor = self.origin;
    if (point.x != CGFLOAT_MAX) anchor.x = self.origin.x + point.x;
    if (point.y != CGFLOAT_MAX) anchor.y = self.origin.y + point.y;
    
    PhysicsSpringBehavior *springBehavior = [[PhysicsSpringBehavior alloc] initWithTarget:self anchorPoint:anchor];
    springBehavior.tension = point.tension;
    springBehavior.haptics = point.haptics;
    springBehavior.influence = [self influenceAreaFromPoint:point];
    [self.animator addBehavior:springBehavior];
    
    if (point.damping > 0.0)
    {
        PhysicsFrictionBehavior *frictionBehavior = [[PhysicsFrictionBehavior alloc] initWithTarget:self];
        frictionBehavior.friction = point.damping;
        if (springBehavior.influence) frictionBehavior.influence = springBehavior.influence;
        [self.animator addBehavior:frictionBehavior];
    }
}

- (void)addConstantGravityBehavior:(InteractablePoint*)point
{
    CGPoint anchor = self.origin;
    if (point.x != CGFLOAT_MAX) anchor.x = self.origin.x + point.x;
    if (point.y != CGFLOAT_MAX) anchor.y = self.origin.y + point.y;
    
    PhysicsGravityWellBehavior *gravityBehavior = [[PhysicsGravityWellBehavior alloc] initWithTarget:self anchorPoint:anchor];
    gravityBehavior.strength = point.strength;
    gravityBehavior.falloff = point.falloff;
    gravityBehavior.influence = [self influenceAreaFromPoint:point];
    [self.animator addBehavior:gravityBehavior];
    
    if (point.damping > 0.0)
    {
        PhysicsFrictionBehavior *frictionBehavior = [[PhysicsFrictionBehavior alloc] initWithTarget:self];
        frictionBehavior.friction = point.damping;
        if (gravityBehavior.strength > 0.0) frictionBehavior.haptics = point.haptics;
        if (gravityBehavior.influence) frictionBehavior.influence = gravityBehavior.influence;
        else frictionBehavior.influence = [self influenceAreaWithRadius:1.4 * point.falloff fromAnchor:anchor];
        [self.animator addBehavior:frictionBehavior];
    }
}

- (void)addConstantFrictionBehavior:(InteractablePoint*)point
{
    if (point.damping > 0.0)
    {
        PhysicsFrictionBehavior *frictionBehavior = [[PhysicsFrictionBehavior alloc] initWithTarget:self];
        frictionBehavior.friction = point.damping;
        frictionBehavior.haptics = point.haptics;
        frictionBehavior.influence = [self influenceAreaFromPoint:point];
        [self.animator addBehavior:frictionBehavior];
    }
}

- (PhysicsArea*)influenceAreaFromPoint:(InteractablePoint*)point
{
    if (!point.limitX && !point.limitY) return nil;
    CGPoint minPoint = CGPointMake(-CGFLOAT_MAX, -CGFLOAT_MAX);
    CGPoint maxPoint = CGPointMake(CGFLOAT_MAX, CGFLOAT_MAX);
    if (point.limitX && point.limitX.min != -CGFLOAT_MAX) minPoint.x = self.origin.x + point.limitX.min;
    if (point.limitX && point.limitX.max != CGFLOAT_MAX) maxPoint.x = self.origin.x + point.limitX.max;
    if (point.limitY && point.limitY.min != -CGFLOAT_MAX) minPoint.y = self.origin.y + point.limitY.min;
    if (point.limitY && point.limitY.max != CGFLOAT_MAX) maxPoint.y = self.origin.y + point.limitY.max;
    return [[PhysicsArea alloc] initWithMinPoint:minPoint maxPoint:maxPoint];
}

- (PhysicsArea*)influenceAreaWithRadius:(CGFloat)radius fromAnchor:(CGPoint)anchor
{
    if (radius <= 0.0) return nil;
    CGPoint minPoint = CGPointMake(anchor.x - radius, anchor.y - radius);
    CGPoint maxPoint = CGPointMake(anchor.y + radius, anchor.y + radius);
    return [[PhysicsArea alloc] initWithMinPoint:minPoint maxPoint:maxPoint];
}

@end

File: android/interactable/src/main/java/com/wix/interactable/InteractableView.java
package com.wix.interactable;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.PointF;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ViewGroup;

import com.wix.interactable.physics.PhysicsAnchorBehavior;
import com.wix.interactable.physics.PhysicsAnimator;
import com.wix.interactable.physics.PhysicsArea;
import com.wix.interactable.physics.PhysicsBehavior;
import com.wix.interactable.physics.PhysicsBounceBehavior;
import com.wix.interactable.physics.PhysicsFrictionBehavior;
import com.wix.interactable.physics.PhysicsGravityWellBehavior;
import com.wix.interactable.physics.PhysicsSpringBehavior;

import java.util.ArrayList;

public class InteractableView extends ViewGroup implements PhysicsAnimator.PhysicsAnimatorListener {

    private boolean originSet;
    private PointF origin;
    private PhysicsAnimator animator;
    private PhysicsBehavior dragBehavior;

    private PointF dragStartCenter;
    private PointF dragStartLocation;
    private PointF dragLastLocation;
    private boolean initialPositionSet;

    private boolean verticalOnly;
    private boolean horizontalOnly;
    private PointF initialPosition;

    private InteractableArea boundaries;
    private InteractableSpring dragWithSprings;
    private float dragToss;
    private ArrayList<InteractablePoint> snapPoints = new ArrayList<>();
    private ArrayList<InteractablePoint> springPoints = new ArrayList<>();
    private ArrayList<InteractablePoint> gravityPoints = new ArrayList<>();

    private InteractionListener listener;


    public InteractableView(Context context) {
        super(context);
        init();
    }

    public InteractableView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public InteractableView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public InteractableView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    private void init() {
        originSet = false;
        initialPositionSet = false;
        initializeAnimator();
    }

    public void setEventListener(InteractionListener listener) {
        this.listener = listener;
    }

    @Override
    public void onAnimatorPause() {

    }

    @Override
    public void onAnimationFrame() {
        listener.onAnimatedEvent(getTranslationX(), getTranslationY());
    }

    private void initializeAnimator() {
        animator = new PhysicsAnimator(this);
        animator.setListener(this);
    }

    private PointF getCurrentPosition() {
        return new PointF(getTranslationX(),getTranslationY());
    }

    private PhysicsBehavior addTempDragBehavior(InteractableSpring drag) {
        PhysicsBehavior res = null;

        if (drag == null || drag.tension == Float.MAX_VALUE) {
            PhysicsAnchorBehavior anchorBehavior = new PhysicsAnchorBehavior(this, getCurrentPosition());
            res = anchorBehavior;
            this.animator.addTempBehavior(anchorBehavior);
        }
        else {
            PhysicsSpringBehavior springBehavior = new PhysicsSpringBehavior(this, getCurrentPosition());
            springBehavior.tension = drag.tension;
            res = springBehavior;
            this.animator.addTempBehavior(springBehavior);
        }
        if (drag != null && drag.damping > 0.0)
        {
            PhysicsFrictionBehavior frictionBehavior = new PhysicsFrictionBehavior(this,drag.damping);
            this.animator.addTempBehavior(frictionBehavior);
        }

        return res;
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        Log.d("InteractableView","onInterceptTouchEvent action = " + ev.getAction());
        return true;
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {

    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {

//        Log.d("InteractableView","onTouchEvent action = " + event.getAction());

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                this.animator.removeTempBehaviors();
                this.animator.setDragging(true);
                this.dragStartLocation = new PointF(event.getX(),event.getY());
                this.dragBehavior = addTempDragBehavior(this.dragWithSprings);


                break;

            case MotionEvent.ACTION_MOVE:

                float newX = getTranslationX() + event.getX() - dragStartLocation.x;
                float newY = getTranslationY() + event.getY() - dragStartLocation.y;
                if (horizontalOnly) {
                    newY = 0;
                }
                if (verticalOnly) {
                    newX = 0;
                }

                this.dragBehavior.setAnchorPoint(new PointF(newX,newY));
//                float dx = event.getX() - dragLastLocation.x;
//                float dy = event.getY() - dragLastLocation.y;
//                if (horizontalOnly) {
//                    dy = 0;
//                }
//                if (verticalOnly) {
//                    dx = 0;
//                }
//
//                this.dragBehavior.moveAnchorPoint(dx,dy);

                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                handleEndOfDrag();
                break;

        }
        this.dragLastLocation = new PointF(event.getX(),event.getY());

        return true;
    }

    private void handleEndOfDrag() {
        this.animator.removeTempBehaviors();
        this.animator.setDragging(false);

        PointF velocity = this.animator.getTargetVelocity(this);
        if (this.horizontalOnly) velocity.y = 0;
        if (this.verticalOnly) velocity.x = 0;
        float toss = 0.1f;
        if (this.dragWithSprings != null) toss = this.dragWithSprings.toss;

        Log.d("InteractableView","handleEndOfDrag velocity = " + velocity);

        PointF projectedCenter = new PointF(getTranslationX() + toss*velocity.x,
                                            getTranslationY() + toss*velocity.y);

        InteractablePoint snapPoint = InteractablePoint.findClosestPoint(snapPoints,projectedCenter);

        addTempSnapToPointBehavior(snapPoint);
        addTempBounceBehaviorWithBoundaries(this.boundaries);

    }

    private void addTempSnapToPointBehavior(InteractablePoint snapPoint) {
        if (snapPoint == null) {
            return;
        }
        listener.onSnap(snapPoints.indexOf(snapPoint), snapPoint.id);
        PhysicsSpringBehavior snapBehavior = new PhysicsSpringBehavior(this,snapPoint.positionWithOrigin());
        snapBehavior.tension = snapPoint.tension;

        this.animator.addTempBehavior(snapBehavior);

        float damping = 0.7f;
        if (snapPoint.damping > 0.0) damping = snapPoint.damping;
        PhysicsFrictionBehavior frictionBehavior = new PhysicsFrictionBehavior(this,damping);
        this.animator.addTempBehavior(frictionBehavior);
        //TODO - continue here!

    }

    private void addTempBounceBehaviorWithBoundaries(InteractableArea boundaries) {
        Log.d("InteractableView","addTempBounceBehaviorWithBoundaries influenceArea = " + boundaries);
//        if (boundaries != null && boundaries.getBounce() > 0.0)
        if (boundaries != null)
        {
            PointF minPoint = new PointF(-Float.MAX_VALUE,-Float.MAX_VALUE);
            if (boundaries.getLeft() != -Float.MAX_VALUE) minPoint.x = boundaries.getLeft();
            if (boundaries.getTop() != -Float.MAX_VALUE) minPoint.y = boundaries.getTop();

            PointF maxPoint = new PointF(Float.MAX_VALUE, Float.MAX_VALUE);
            if (boundaries.getRight() != Float.MAX_VALUE) maxPoint.x = boundaries.getRight();
            if (boundaries.getBottom() != Float.MAX_VALUE) maxPoint.y = boundaries.getBottom();

            PhysicsBounceBehavior bounceBehavior = new PhysicsBounceBehavior(this,minPoint, maxPoint, boundaries.getBounce(), boundaries.isHaptic());
            this.animator.addTempBehavior(bounceBehavior);
//            this.animator.addBehavior(bounceBehavior);
        }
    }
    private void addConstantBoundaries(InteractableArea boundaries) {
        Log.d("InteractableView","addTempBounceBehaviorWithBoundaries influenceArea = " + boundaries);
//        if (boundaries != null && boundaries.getBounce() > 0.0)
        if (boundaries != null)
        {
            PointF minPoint = new PointF(-Float.MAX_VALUE,-Float.MAX_VALUE);
            if (boundaries.getLeft() != -Float.MAX_VALUE) minPoint.x = boundaries.getLeft();
            if (boundaries.getTop() != -Float.MAX_VALUE) minPoint.y = boundaries.getTop();

            PointF maxPoint = new PointF(Float.MAX_VALUE, Float.MAX_VALUE);
            if (boundaries.getRight() != Float.MAX_VALUE) maxPoint.x = boundaries.getRight();
            if (boundaries.getBottom() != Float.MAX_VALUE) maxPoint.y = boundaries.getBottom();

            PhysicsBounceBehavior bounceBehavior = new PhysicsBounceBehavior(this,minPoint, maxPoint,0, boundaries.isHaptic());

            this.animator.addBehavior(bounceBehavior);
        }
    }

    private void addConstantSpringBehavior(InteractablePoint point) {
        PointF anchor = new PointF(0,0);
        if (point.x != Float.MAX_VALUE) anchor.x = point.x;
        if (point.y != Float.MAX_VALUE) anchor.y = point.y;

        PhysicsSpringBehavior springBehavior = new PhysicsSpringBehavior(this,anchor);
        springBehavior.tension = point.tension;
        springBehavior.setInfluence(influenceAreaFromPoint(point));

        this.animator.addBehavior(springBehavior);

        if (point.damping > 0.0) {
            PhysicsFrictionBehavior frictionBehavior = new PhysicsFrictionBehavior(this,point.damping);
            frictionBehavior.setInfluence(influenceAreaFromPoint(point));

            this.animator.addBehavior(frictionBehavior);
        }
    }

    private void addConstantGravityBehavior(InteractablePoint point) {
        PointF anchor = new PointF(0,0);
        if (point.x != Float.MAX_VALUE) anchor.x = point.x;
        if (point.y != Float.MAX_VALUE) anchor.y = point.y;

        PhysicsGravityWellBehavior gravityBehavior = new PhysicsGravityWellBehavior(this,anchor);
        gravityBehavior.setStrength(point.strength);
        gravityBehavior.setFalloff(point.falloff );
        PhysicsArea influenceArea = influenceAreaFromPoint(point);
        gravityBehavior.setInfluence(influenceArea);

        this.animator.addBehavior(gravityBehavior);

        if (point.damping > 0.0) {
            PhysicsFrictionBehavior frictionBehavior = new PhysicsFrictionBehavior(this,point.damping);

            if (influenceArea == null) {
                frictionBehavior.setInfluence(influenceAreaWithRadius(1.4f * point.falloff,anchor));
            }
            else {
                frictionBehavior.setInfluence(influenceArea);
            }
            this.animator.addBehavior(frictionBehavior);
        }
    }

    private void addConstantFrictionBehavior(InteractablePoint point) {
        PhysicsFrictionBehavior frictionBehavior = new PhysicsFrictionBehavior(this,point.damping);
        frictionBehavior.setInfluence(influenceAreaFromPoint(point));
        this.animator.addBehavior(frictionBehavior);
    }

    private PhysicsArea influenceAreaFromPoint(InteractablePoint point)
    {
        if (point.influenceArea == null) return null;
        
        PointF minPoint = new PointF(-Float.MAX_VALUE, -Float.MAX_VALUE);
        PointF maxPoint = new PointF(Float.MAX_VALUE, Float.MAX_VALUE);

        if (point.influenceArea.getLeft() != -Float.MAX_VALUE) minPoint.x = point.influenceArea.getLeft();
        if (point.influenceArea.getRight() != Float.MAX_VALUE) maxPoint.x = point.influenceArea.getRight();
        if (point.influenceArea.getTop() != -Float.MAX_VALUE) minPoint.y = point.influenceArea.getTop();
        if (point.influenceArea.getBottom() != Float.MAX_VALUE) maxPoint.y = point.influenceArea.getBottom();

        return new PhysicsArea(minPoint,maxPoint);
    }

    private PhysicsArea influenceAreaWithRadius(float radius,PointF anchor) {
        if (radius <= 0.0) return null;
        PointF minPoint = new PointF(anchor.x - radius, anchor.y - radius);
        PointF maxPoint = new PointF(anchor.x + radius, anchor.y + radius);
        Log.d("InteractableView","influenceAreaWithRadius minPoint = " + minPoint + " maxPoint = " + maxPoint);
        return new PhysicsArea(minPoint,maxPoint);
    }

    public void setVerticalOnly(boolean verticalOnly) {
        this.verticalOnly = verticalOnly;
    }

    public void setHorizontalOnly(boolean horizontalOnly) {
        this.horizontalOnly = horizontalOnly;
    }

    public void setInitialPosition(PointF initialPosition) {
        this.initialPosition = initialPosition;
        setTranslationX(initialPosition.x);
        setTranslationY(initialPosition.y);
    }

    public void setBoundaries(InteractableArea boundaries) {
        this.boundaries = boundaries;
        addConstantBoundaries(boundaries);
    }


    public void setDragWithSprings(InteractableSpring dragWithSprings) {
        this.dragWithSprings = dragWithSprings;
    }

    public void setDragToss(float dragToss) {
        this.dragToss = dragToss;
    }
    
    public void setSnapPoints(ArrayList snapPoints) {
        this.snapPoints = snapPoints;
    }

    public void setSpringsPoints(ArrayList<InteractablePoint> springPoints) {
        this.springPoints = springPoints;
        for (InteractablePoint point : springPoints) {
            addConstantSpringBehavior(point);
        }
    }

    public void setGravityPoints(ArrayList<InteractablePoint> gravityPoints) {
        this.gravityPoints = gravityPoints;
        for (InteractablePoint point : gravityPoints) {
            Log.d("InteractableView","setGravityPoints strength = " + point.strength);
            Log.d("InteractableView","setGravityPoints damping = " + point.damping);

            addConstantGravityBehavior(point);
        }
    }

    public void setFrictionAreas(ArrayList<InteractablePoint> frictionAreas) {
        this.gravityPoints = gravityPoints;
        for (InteractablePoint point : frictionAreas) {
            Log.d("InteractableView","setFrictionAreas damping = " + point.damping);
            addConstantFrictionBehavior(point);
        }
    }

    public interface InteractionListener {
        void onSnap(int indexOfSnapPoint, String snapPointId);
        void onAnimatedEvent(float x, float y);
    }
}
